### ISSUE TITLE
XGrid controlled Selection does not unselect on checkbox click

### ISSUE THREAD

**Issue by Fibs7000** [mui-org/material-ui-x#1577]
When using XGrid with controlled selection selecting works with the checkboxes but unselecting does not.

- [x] The issue is present in the latest release.
- [ ] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

When clicking on a selected checkbox, the line stays selected and a console output says that selectionModelChange is triggerd but not changed.

## Expected Behavior 🤔

When clicking on a selected checkbox the line should be unselected.

#### Minimal example

```typescript
import { XGrid } from "@material-ui/x-grid";
import React, { useState } from "react";

const columns = [
  {
    field: "name",
    width: 300,
    headerName: "Groupname",
  },
];

const rows = [
  {name: "alice", id: "1"},
  {name: "bob", id: "2"},
  {name: "mark", id: "3"},
]

export function Example() {
  const [gridSelection, setGridSelection] = useState<(string | number)[]>([]);

  return (
      <XGrid
        headerHeight={40}
        disableColumnSelector
        selectionModel={gridSelection}
        onSelectionModelChange={(newSelection) => {
          console.log(newSelection);
          setGridSelection(newSelection.selectionModel);
        }}
        checkboxSelection
        columns={columns}
        rows={rows}
      />
  );
}


```

Steps:

1. Click on alice checkbox
2. Click again
3. Row is not unselected

## Context 🔦

Does not work with both selectionModel set or unset.

## Your Environment 🌎

<details>
  <summary>`npx @material-ui/envinfo`</summary>
  
```
  System:
    OS: Linux 5.11 Pop!_OS 20.10
    CPU: (16) x64 AMD Ryzen 7 5800X 8-Core Processor
    Memory: 1.19 GB / 31.34 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 15.7.0 - ~/.nvm/versions/node/v15.7.0/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v15.7.0/bin/yarn
    npm: 7.4.3 - ~/.nvm/versions/node/v15.7.0/bin/npm
    Watchman: 4.9.0 - /usr/bin/watchman
  Managers:
    Apt: 2.1.10 - /usr/bin/apt
    Composer: 2.0.9 - /usr/local/bin/composer
    Gradle: 4.4.1 - /usr/bin/gradle
    RubyGems: 3.1.2 - /usr/bin/gem
  Utilities:
    Make: 4.3 - /usr/bin/make
    GCC: 10.2.0 - /usr/bin/gcc
    Git: 2.27.0 - /usr/bin/git
  Virtualization:
    Docker: 20.10.6 - /usr/bin/docker
  SDKs:
    Android SDK:
      API Levels: 23, 27, 28, 29, 30
      Build Tools: 28.0.3, 29.0.2, 30.0.3
      System Images: android-30 | Google APIs Intel x86 Atom
  IDEs:
    Nano: 5.2 - /usr/bin/nano
    VSCode: 1.55.2 - /usr/bin/code
  Languages:
    Bash: 5.0.17 - /usr/bin/bash
    Java: 11.0.11 - /usr/bin/javac
    Perl: 5.30.3 - /usr/bin/perl
    PHP: 7.4.9 - /usr/bin/php
    Python3: 3.8.6 - /usr/bin/python3
    R: 4.0.2 - /usr/bin/R
    Ruby: 2.7.1 - /usr/bin/ruby
  Databases:
    SQLite: 3.33.0 - /usr/bin/sqlite3
  Browsers:
    Chrome: 90.0.4430.93
    Firefox: 88.0
```
</details>

## Order id 💳
 #24533

**Comment by oliviertassinari** [comment]
Please provide a **minimal** reproduction test case with the latest version. This would help a lot 👷 .
A live example would be perfect. [This **codesandbox.io** template](https://material-ui.com/r/issue-template) _may_ be a good starting point. Thank you!

**Comment by Fibs7000** [comment]
https://codesandbox.io/s/interesting-allen-68o26?file=/src/Demo.tsx

Here is a Demo reproducing that bug

**Comment by oliviertassinari** [comment]
@Fibs7000 Thanks, so the issue is related to the version of Material-UI. Haha, yes, I'm actually surprised to see 

https://github.com/mui-org/material-ui-x/blob/3e4be4de6cd41030eee75e8485c64a9a5ba48502/packages/grid/_modules_/grid/components/columnSelection/GridHeaderCheckbox.tsx#L35

We have stopped documenting this signature a long time ago.

**Comment by oliviertassinari** [comment]
The fix is trivial: https://github.com/mui-org/material-ui/pull/25871

**Comment by HeavenlyEntity** [comment]
Same exact behavior on the `DataGridPro` component @ZeeshanTamboli if you enable `disableSelectionOnClick` the checkbox does not trigger a selection. Nor trigger an event.

**Comment by HeavenlyEntity** [comment]
> @HeavenlyEntity could you open another issue with a reproduction? The demo in https://mui.com/components/data-grid/demo/ uses `disableSelectionOnClick` and the prop is being called.

Before I put in an issue what is the version of the grid currently at?

**Comment by m4theushw** [comment]
@HeavenlyEntity v5.1.0.

**Comment by HeavenlyEntity** [comment]
> @HeavenlyEntity v5.1.0.

👍 right on, mine is at `5.0.1` so I am going to test updating the package and rerun.

**Comment by HeavenlyEntity** [comment]
> > @HeavenlyEntity v5.1.0.
> 
> 👍 right on, mine is at `5.0.1` so I am going to test updating the package and rerun.

Yeah still not working I'll put in an issue. But so you guys see whats going on. 

![mPYpm5OhJd](https://user-images.githubusercontent.com/39228141/144627664-5cae1575-ad67-480c-9d83-fa2fbb0847db.gif)

**Comment by rave** [comment]
@HeavenlyEntity 
I had the same issue after migrating to new MUI. Fixed that by ensuring my ids that i pass to `selectionModel` prop have Number type, not string.

**Comment by Adrian-Benavente** [comment]
I have the exact same problem as @HeavenlyEntity and @rave's solution didn't work for me

**Comment by Adrian-Benavente** [comment]
Ok, I'll show you what was happening in my case. Hopefully it helps other people. In my case, it was another library stepping in the way of MUI styles as you can see in this screenshot
![image](https://user-images.githubusercontent.com/11258517/149185352-000f1d21-920c-4adb-a05a-ec22fc233434.png)

## The solution:
Creating a CSS rule for the MUI checkbox element that has more precedence; not the ideal soution, but necessary if you need to keep the other library.
```
.PrivateSwitchBase-input {
    left: initial !important;
}
```

**Comment by HeavenlyEntity** [comment]
> @HeavenlyEntity I had the same issue after migrating to new MUI. Fixed that by ensuring my ids that i pass to `selectionModel` prop have Number type, not string.

Hey @rave I was able to fix the problem after removing the pro version of another UI library (`mdbreact`) that was overriding the click event. I was able to fix the issue with the theming by upgrading to the latest version of the package. If you have that library it will break the checkbox no doubt.

**Comment by werty0u** [comment]
Was there ever a fix for this issue since this its closed? I am on V 5.4 of @mui/material and @mui/x-data-grid and am still experiencing this behavior when using `checkboxSelection`. I was using strings as rowIds and tried switching to numbers to see if that fixed things but it didn't. Clicking on the select checkbox doesnt trigger the onSlectionChange at all.

**Comment by raphaelbadia** [comment]
It happened in my codebase, was caused by a developer that somewhy resized all the inputs to a very small square, in a global file.
```css
// bad global code
input[type='checkbox'] {
  width: 15px; // culprit
  height: 15px; // culprit
  appearance: none;
  margin-right: 10px;
  background-color: white;
  outline: 0;
  border: 0;
  border-radius: 4px;
  box-shadow: none !important;
}
```
<img width="304" alt="CleanShot 2022-05-06 at 22 42 26@2x" src="https://user-images.githubusercontent.com/1901628/167214002-e6835548-f76d-4134-9d7c-6dd4257e691e.png">


Sometimes, it's not the lib... 😬💀
